Installation issue: DCMTK with TIFF support

### Steps to reproduce

Trying to build dcmtk with tiff support fails during the linking stage.  While asking for help in the slack channel I was informed that the issue is the libtiff recipe needs to be updated to include a `libs()` method simliar to libpng.

### Error message

Building dcmtk with tiff support should work.

### Information on your system

* **Spack:** 0.23.0.dev0 (e5f5749d676d0cc4e247dd41c73bc31ff6b6945e)
* **Python:** 3.9.17
* **Platform:** linux-ubuntu16.04-ivybridge
* **Concretizer:** clingo


### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

My original issue was building `dcmtk` with tiff support, in Slack it was suggested the issue was due to the missing `libs()` method.

The error when building dcmtk is:

```
[ 71%] Linking CXX executable ../../bin/dcm2pnm
cd /bigtmp/tmp/moloney/spack-stage/spack-stage-dcmtk-3.6.7-vnlggacdmss5wdx4hy4wndka2eapd3fw/spack-build-vnlggac/dcmimage/apps && /usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-12.3.0/cmake-3.29.4-di2xxf45fr624obgrzeqbymb7eusi6ym/bin/cmake -E cmake_link_script CMakeFiles/dcm2pnm.dir/link.txt --verbose=1
/usr/contrib/byoe/repo/._internal/spack/lib/spack/env/gcc/g++  -D_XOPEN_SOURCE_EXTENDED -D_XOPEN_SOURCE=500 -D_BSD_SOURCE -D_DEFAULT_SOURCE -D_BSD_COMPAT -D_OSF_SOURCE -D_POSIX_C_SOURCE=199506L -O3 -DNDEBUG CMakeFiles/dcm2pnm.dir/dcm2pnm.cc.o -o ../../bin/dcm2pnm  -Wl,-rpath,/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-12.3.0/zlib-ng-2.1.6-uqbferhpitbgl3cgko4gcq2vly7wmjrk/lib:/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-12.3.0/libxml2-2.10.3-kb6x5eijzqczhjonnipzgkoui3ss5u5k/lib:/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-12.3.0/libiconv-1.17-fmk4tnaxichiitzhpladhukugxxjvpjv/lib:/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-12.3.0/libjpeg-turbo-3.0.0-6fvwoez4sgareo4f7wc3c6vasbwgjrhl/lib:/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-12.3.0/libpng-1.6.39-klgrztcrtd643w7jmxayhieta5w66hei/lib:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: ../../lib/libdcmimage.a ../../lib/libdcmimgle.a ../../lib/libdcmdata.a ../../lib/liboflog.a ../../lib/libofstd.a /usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-12.3.0/zlib-ng-2.1.6-uqbferhpitbgl3cgko4gcq2vly7wmjrk/lib/libz.so /usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-12.3.0/libxml2-2.10.3-kb6x5eijzqczhjonnipzgkoui3ss5u5k/lib/libxml2.so /usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-12.3.0/libiconv-1.17-fmk4tnaxichiitzhpladhukugxxjvpjv/lib/libiconv.so /usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-12.3.0/libiconv-1.17-fmk4tnaxichiitzhpladhukugxxjvpjv/lib/libcharset.so -lnsl -lpthread /usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-12.3.0/libjpeg-turbo-3.0.0-6fvwoez4sgareo4f7wc3c6vasbwgjrhl/lib/libjpeg.so /usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-12.3.0/libpng-1.6.39-klgrztcrtd643w7jmxayhieta5w66hei/lib/libpng.so
/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-5.4.0/binutils-2.41-fxgfdiqsjh2bp3y5enqq4nzberrctbc3/bin/ld: ../../lib/libdcmimage.a(dipitiff.cc.o): in function `DiTIFFPlugin::write(DiImage*, _IO_FILE*, unsigned long) const':
dipitiff.cc:(.text+0x179): undefined reference to `TIFFFdOpen'
/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-5.4.0/binutils-2.41-fxgfdiqsjh2bp3y5enqq4nzberrctbc3/bin/ld: dipitiff.cc:(.text+0x198): undefined reference to `TIFFSetField'
/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-5.4.0/binutils-2.41-fxgfdiqsjh2bp3y5enqq4nzberrctbc3/bin/ld: dipitiff.cc:(.text+0x1ab): undefined reference to `TIFFSetField'
/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-5.4.0/binutils-2.41-fxgfdiqsjh2bp3y5enqq4nzberrctbc3/bin/ld: dipitiff.cc:(.text+0x1bf): undefined reference to `TIFFSetField'
/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-5.4.0/binutils-2.41-fxgfdiqsjh2bp3y5enqq4nzberrctbc3/bin/ld: dipitiff.cc:(.text+0x1d3): undefined reference to `TIFFSetField'
/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-5.4.0/binutils-2.41-fxgfdiqsjh2bp3y5enqq4nzberrctbc3/bin/ld: dipitiff.cc:(.text+0x1e6): undefined reference to `TIFFSetField'
/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-5.4.0/binutils-2.41-fxgfdiqsjh2bp3y5enqq4nzberrctbc3/bin/ld: ../../lib/libdcmimage.a(dipitiff.cc.o):dipitiff.cc:(.text+0x206): more undefined references to `TIFFSetField' follow
/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-5.4.0/binutils-2.41-fxgfdiqsjh2bp3y5enqq4nzberrctbc3/bin/ld: ../../lib/libdcmimage.a(dipitiff.cc.o): in function `DiTIFFPlugin::write(DiImage*, _IO_FILE*, unsigned long) const':
dipitiff.cc:(.text+0x2b7): undefined reference to `TIFFWriteScanline'
/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-5.4.0/binutils-2.41-fxgfdiqsjh2bp3y5enqq4nzberrctbc3/bin/ld: dipitiff.cc:(.text+0x2da): undefined reference to `TIFFFlushData'
/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-5.4.0/binutils-2.41-fxgfdiqsjh2bp3y5enqq4nzberrctbc3/bin/ld: dipitiff.cc:(.text+0x2e2): undefined reference to `TIFFCleanup'
/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-5.4.0/binutils-2.41-fxgfdiqsjh2bp3y5enqq4nzberrctbc3/bin/ld: dipitiff.cc:(.text+0x356): undefined reference to `TIFFSetField'
/usr/contrib/byoe/repo/._internal/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder/linux-ubuntu16.04-ivybridge/gcc-5.4.0/binutils-2.41-fxgfdiqsjh2bp3y5enqq4nzberrctbc3/bin/ld: ../../lib/libdcmimage.a(dipitiff.cc.o): in function `DiTIFFPlugin::getLibraryVersionString[abi:cxx11]()':
dipitiff.cc:(.text+0x3e2): undefined reference to `TIFFGetVersion'
collect2: error: ld returned 1 exit status
make[2]: *** [dcmimage/apps/CMakeFiles/dcm2pnm.dir/build.make:111: bin/dcm2pnm] Error 1
```

--- a/var/spack/repos/builtin/packages/dcmtk/package.py
+++ b/var/spack/repos/builtin/packages/dcmtk/package.py
@@ -50,6 +50,8 @@ class Dcmtk(CMakePackage):
 
     conflicts("platform=darwin target=aarch64:", when="@:3.6.6")
 
+    patch("tiff-3.6.7.patch", when="@3.6.7")
+
     def patch(self):
         # Backport 3.6.4
         if self.spec.satisfies("@:3.6.3 %fj"):

--- a/var/spack/repos/builtin/packages/dcmtk/tiff-3.6.7.patch
+++ b/var/spack/repos/builtin/packages/dcmtk/tiff-3.6.7.patch
@@ -1,12 +1,12 @@
 diff --color=auto --color=never -Naur a/CMake/3rdparty.cmake b/CMake/3rdparty.cmake
 --- a/CMake/3rdparty.cmake	2022-04-28 15:47:25
-+++ b/CMake/3rdparty.cmake	2024-07-12 14:55:11
++++ b/CMake/3rdparty.cmake	2024-07-12 15:04:19
 @@ -38,7 +38,7 @@
          message(STATUS "Info: DCMTK TIFF support will be enabled")
          include_directories(${TIFF_INCLUDE_DIR} ${JPEG_INCLUDE_DIR})
        endif()
 -      set(LIBTIFF_LIBS ${TIFF_LIBRARY} ${TIFF_EXTRA_LIBS_STATIC} ${JPEG_LIBRARY})
-+      set(LIBTIFF_LIBS ${TIFF_LIBRARIES} ${TIFF_EXTRA_LIBS_STATIC} ${JPEG_LIBRARY})
++      set(LIBTIFF_LIBS ${TIFF_LIBRARIES} ${TIFF_EXTRA_LIBS_STATIC} ${JPEG_LIBRARIES})
      endif()
    endif()
  

--- a/var/spack/repos/builtin/packages/dcmtk/tiff-3.6.7.patch
+++ b/var/spack/repos/builtin/packages/dcmtk/tiff-3.6.7.patch
@@ -1,0 +1,12 @@
+diff --color=auto --color=never -Naur a/CMake/3rdparty.cmake b/CMake/3rdparty.cmake
+--- a/CMake/3rdparty.cmake	2022-04-28 15:47:25
++++ b/CMake/3rdparty.cmake	2024-07-12 14:55:11
+@@ -38,7 +38,7 @@
+         message(STATUS "Info: DCMTK TIFF support will be enabled")
+         include_directories(${TIFF_INCLUDE_DIR} ${JPEG_INCLUDE_DIR})
+       endif()
+-      set(LIBTIFF_LIBS ${TIFF_LIBRARY} ${TIFF_EXTRA_LIBS_STATIC} ${JPEG_LIBRARY})
++      set(LIBTIFF_LIBS ${TIFF_LIBRARIES} ${TIFF_EXTRA_LIBS_STATIC} ${JPEG_LIBRARY})
+     endif()
+   endif()
+ 

